kamon-core/src/main/scala/kamon/ContextStorage.scala

```diff
@@ -0,0 +1,47 @@
+package kamon
+
+import kamon.context.{Context, Storage}
+import kamon.trace.Span
+
+trait ContextStorage {
+  private val _contextStorage = Storage.ThreadLocal()
+
+  def currentContext(): Context =
+    _contextStorage.current()
+
+  def currentSpan(): Span =
+    _contextStorage.current().get(Span.ContextKey)
+
+  def storeContext(context: Context): Storage.Scope =
+    _contextStorage.store(context)
+
+  def withContext[T](context: Context)(f: => T): T = {
+    val scope = _contextStorage.store(context)
+    try {
+      f
+    } finally {
+      scope.close()
+    }
+  }
+
+  def withContextKey[T, K](key: Context.Key[K], value: K)(f: => T): T =
+    withContext(currentContext().withKey(key, value))(f)
+
+  def withSpan[T](span: Span)(f: => T): T =
+    withSpan(span, true)(f)
+
+  def withSpan[T](span: Span, finishSpan: Boolean)(f: => T): T = {
```

Comment:
I think is a good idea we have a variant of withSpan like this:

Reply:
I'm not entirely sure about it :thinking:, the only reason for catching the exceptions on the `withSpan` variant would be to add it to the Span, not to introduce any opinion on how the actual error handling.

I would vote for leaving that to the users.

Reply:
In this case I agree 100% with not introduce any opinion on how should manage the actual error handling. I think that is a responsibility of user.